server/broadcast.go

```diff
@@ -22,56 +22,155 @@ import (
 	"github.com/livepeer/lpms/stream"
 )
 
-func selectOrchestrator(n *core.LivepeerNode, cpl core.PlaylistManager) (*BroadcastSession, error) {
+type BroadcastSessionsManager struct {
+	sessList []*BroadcastSession
+	sessMap  map[string]*BroadcastSession
+	sessLock *sync.Mutex
+	minOrchs int
 
+	refreshing bool // only allow one refresh in-flight
+	finished   bool // set at stream end
+
+	createSessions func() ([]*BroadcastSession, error)
+}
+
+func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+	numSess := len(bsm.sessList)
+
+	if numSess < bsm.minOrchs {
```

Comment:
How we handle minOrchs is a bit problematic right now. I think this would cause a refresh to be fired every time an O fails or if the number of Os in the pool is less than `minOrchs` (4 right now?) - this would be common during offchain scenarios and testing.

Also, since 4 is the maximum number of inflight sessions, we want to give ourselves some more overhead there, so let's say 8.

We should fix this with something like `minOrchs=min(OrchestratorPool.Size(), HTTPTimeout / SegLen * 2)`, and set the refresh threshold to something like `refreshThreshold=ceil(minOrchs/2)`. This would allow the refresh threshold to scale along with the pool size up to 8 orchestrators.

O Pool Size | Quantity to Refresh (minOrchs) | Refresh Below (numSessions < refreshThreshold)
-- | -- | --
1 | 1 | 1
2 | 2 | 1
3 | 3 | 2
4 | 4 | 2
5 | 5 | 3
6 | 6 | 3
7 | 7 | 4
8 | 8 | 4
9 | 8 | 4
10 | 8 | 4
11 | 8 | 4
12 | 8 | 4